module/zfs/zfs_ioctl.c

```diff
@@ -2342,26 +2358,44 @@ zfs_ioc_snapshot_list_next(zfs_cmd_t *zc)
 		return (SET_ERROR(ESRCH));
 	}
 
-	error = dmu_snapshot_list_next(os,
-	    sizeof (zc->zc_name) - strlen(zc->zc_name),
-	    zc->zc_name + strlen(zc->zc_name), &zc->zc_obj, &zc->zc_cookie,
-	    NULL);
+	while (error == 0) {
+		error = dmu_snapshot_list_next(os,
+		    sizeof (zc->zc_name) - strlen(zc->zc_name),
+		    zc->zc_name + strlen(zc->zc_name), &zc->zc_obj,
+		    &zc->zc_cookie, NULL);
```

Comment:
This was discussed earlier in the PR: https://github.com/zfsonlinux/zfs/pull/8077#issuecomment-434823796. As I understand a possible way to do it would be to create and use a new channel program interface that allowed named snapshot lookup and iteration